### PR TITLE
[router] Cache auction bids; prevent overbidding by checking current outstanding liquidity

### PR DESCRIPTION
**Pull request by jakekidd**
## Checklist

- [ ] Test manually on `test-ui` using remote chains.
- [ ] Run load tests.
- [x] Update documentation if needed.
- [x] Update CHANGELOG.md.

### CODE REVIEW

File: packages/router/src/adapters/cache/auction.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// TODO: This could use some optimization.
+// - For one, we should be storing via Redis (and move this to adapters) to keep the state persistent across restarts.
+// - Store by transactionId in a nested Map (instead of an Array)?
+// - We could also optimize by comparing user asks to actual funding in their wallet, and only consider
+//   promised liquidity actually outstanding if the user *can* go through with the bid.
+/**
+ * Bid caching helper class to ensure we only store one bid per assetId/chainId combination, clean up expired bids,
+ * and provide a quick calculation endpoint to determine outstanding liquidity.
+ */
+export class AuctionCache {
```
Comment (LayneHaber) [inline packages/router/src/adapters/cache/auction.ts:14]: Wondering the best way to handle the case where we win the bid, and the bid expires before we are able to prepare the receiver. Perhaps there should be a way to increase the expiry or set a `confirmed` flag?

**Review by LayneHaber — COMMENTED** [review]
Looks good, pending tests.

Still need to figure out how to extend the lifetime of a bid when processing `SenderPrepared` transactions

File: packages/router/src/adapters/cache/auction.ts (hunk 1 above)
Comment (jakekidd) [reply packages/router/src/adapters/cache/auction.ts:14]: oh right, like we want to keep the liquidity reserved, right?

File: packages/router/src/bindings/contractReader/index.ts [hunk 2]
```diff
@@ -214,6 +214,20 @@ export const handleSingle = async (
     const preparePayload: PreparePayload = _transaction.payload;
     try {
       logger.info("Preparing receiver", requestContext, methodContext);
+      const didConfirm = cache.auctions.confirmBid(
```
Comment (rhlsthrm) [inline packages/router/src/bindings/contractReader/index.ts:217]: This seems a bit unnecessary? Our signature on a bid is essentially our commitment, so I'm not sure what we gain by confirming that it's in the cache.

**Review by rhlsthrm — CHANGES_REQUESTED** [review]
Great job on this, annoying that we replicated some work, but no worries. We can probably just go with this implementation at first. Pending some answers/clarifications. Also, I think we should add the config on the 150% thing.

File: packages/router/src/bindings/contractReader/index.ts (hunk 2 above)
Comment (jakekidd) [reply packages/router/src/bindings/contractReader/index.ts:217]: You're misreading the purpose of this. It's not for our sake here, it's for the sake of keeping that outstanding liquidity "held" as outstanding until we've confirmed that our prepare has deducted from our overall balance.

File: packages/router/src/adapters/cache/auction.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+   * @param bid - bid to add
+   */
+  public addBid(bid: Omit<CachedBid, "confirmed">) {
+    const key = this.getCachedBidCompositeKey(bid.chainId, bid.assetId);
+    const bids = this.auctions.get(key) ?? new Map<string, CachedBid>();
+    bids.set(bid.transactionId, {
+      ...bid,
+      confirmed: false,
+    });
+    this.auctions.set(key, bids);
```
[the remark this answers is not in the document]
Comment (jakekidd) [reply packages/router/src/adapters/cache/auction.ts:52]: They do expire? Each time we calculate outstanding liquidity we remove expired bids...?

File: packages/router/src/bindings/contractReader/index.ts (hunk 2 above)
Comment (jakekidd) [reply packages/router/src/bindings/contractReader/index.ts:217]: https://github.com/connext/nxtp/pull/700/files/7f1b5df049f918c48c9a8f326788ae93b955dc90#r778491800

Comment (jakekidd) [reply packages/router/src/bindings/contractReader/index.ts:217]: So:
`bid sent ---> bid accepted ---> user prepares ---> router sees user prepared -***-> router prepares and liquidity is deducted from router's pool`
***at some point here, asynchronously, the bid could "expire" but we want the liquidity to remain outstanding until our prepare is complete

Comment (jakekidd) [reply packages/router/src/bindings/contractReader/index.ts:217]: I'll add a clarifying comment

File: packages/router/src/adapters/cache/auction.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhlsthrm) [reply packages/router/src/adapters/cache/auction.ts:52]: Oops, I guess I overlooked this.

**Review by rhlsthrm — APPROVED** [review]

**Comment by rhlsthrm** [comment]
This is merged, but I just realized we should really have a way to turn this functionality off (i.e. accept bids always) just in case something goes wrong. We need to make sure things are properly feature-flagged through config.

**Pull request merged**